src/zerocoin_v3.cpp

```diff
@@ -604,51 +596,64 @@ void CSigmaTxInfo::Complete() {
 CSigmaState::CSigmaState() {
 }
 
-int CSigmaState::AddMint(
+bool CSigmaState::AddMintsToStateAndBlockIndex(
         CBlockIndex *index,
-        const sigma::PublicCoin &pubCoin) {
-    sigma::CoinDenomination denomination = pubCoin.getDenomination();
-
-    if (latestCoinIds[denomination] < 1)
-        latestCoinIds[denomination] = 1;
-    int	mintCoinGroupId = latestCoinIds[denomination];
-
-    // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
-    // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
-    // each block will end up in the same accumulator.
-    SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
-    int coinsPerId = ZC_SPEND_V3_COINSPERID;
-    if (coinGroup.nCoins < coinsPerId // there's still space in the accumulator
-        || coinGroup.lastBlock == index // or we have already placed some coins from current block.
-        ) {
-        if (coinGroup.nCoins++ == 0) {
-            // first group of coins for given denomination
-            assert(coinGroup.firstBlock == nullptr);
-            assert(coinGroup.lastBlock == nullptr);
-
-            coinGroup.firstBlock = coinGroup.lastBlock = index;
-        } else {
-            assert(coinGroup.firstBlock != nullptr);
-            assert(coinGroup.lastBlock != nullptr);
-            if (coinGroup.lastBlock->nHeight > index->nHeight) { 
-                assert(coinGroup.lastBlock->nHeight <= index->nHeight);
-            }
+        const CBlock* pblock) {
+
+    std::unordered_map<sigma::CoinDenomination, size_t> blockDenomCounts;
+    for (const auto& mint : pblock->sigmaTxInfo->mints) {
+        blockDenomCounts[mint.getDenomination()]++;
+    }
+
+    for (const auto& pubCoin : pblock->sigmaTxInfo->mints) {
+        sigma::CoinDenomination denomination = pubCoin.getDenomination();
+
+        if (latestCoinIds[denomination] < 1)
+            latestCoinIds[denomination] = 1;
+        int	mintCoinGroupId = latestCoinIds[denomination];
+
+        // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
+        // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
+        // each block will end up in the same accumulator.
+        SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
+        int coinsPerId = ZC_SPEND_V3_COINSPERID;
+        int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
+        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
```

Comment:
Why is the first part of "if" necessary? Check in "oinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit" checks that all the mints in this block will fit in the same accumulator right? And I believe there must be coinsPerId used, not coinsPerIdLimit.

Reply:
I use 15 000 as soft cap(`coinsPerID`) and 16 000(`coinsPerIdLimit`) as hard cap. 

If before add new mints coin in group is greater than or equal 15000, then new mints should be added to next group. which is first case. `coinGroup.nCoins < coinsPerId`

and if mints in group is less than 15000 but new mints + old mints > 16000, then new mints should also be push to new group. which is second case `coinGroup.nCoins + new mints amount < coinsPerIdLimit`

Reply:
Wait, I did not understand why we need 2 caps any more. We must do the following
1. Know how many coins are already in the last accumulator.
2. Count how many new mints are in the new block.
3. If these mints + number of coins already in the accumulator < 16.000, then put them all into the same accumulator, if not, then close the current accumulator and put them all into the next accumulator. 

This must be the whole logic, no soft cap.

Reply:
I see, that's make sense.